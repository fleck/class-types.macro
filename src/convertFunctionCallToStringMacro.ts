import { createMacro, MacroError } from "babel-plugin-macros";
import * as t from "@babel/types";

export default createMacro(({ references }) => {
  let keepImports = true;

  references.default?.forEach(referencePath => {
    keepImports = false;
    let callExpression: t.CallExpression;

    if (referencePath.parentPath.node.type === "CallExpression") {
      callExpression = referencePath.parentPath.node;
    } else {
      throw new MacroError("ct.macro can only be called as a function");
    }

    const literals = callExpression.arguments.filter(function nonLiterals<T>(
      argument: T
    ): argument is Extract<T, { value: any }> {
      return "value" in argument;
    });

    type Argument = t.CallExpression["arguments"][number];

    const identifiers = callExpression.arguments.filter(function nonIdentifiers<
      T extends Argument
    >(argument: T): argument is Extract<T, { type: "Identifier" }> {
      return "type" in argument && argument.type === "Identifier";
    });

    // If any args aren't literals or identifiers throw...

    let argumentsNode;

    if (referencePath.parentPath.parentPath.type === "JSXExpressionContainer") {
      argumentsNode = referencePath.parentPath.parentPath;
    } else {
      argumentsNode = referencePath.parentPath;
    }

    let toAdd;

    type AstNode = typeof literals[number] | typeof identifiers[number];

    const concat = (nodes: AstNode[], index: number): t.BinaryExpression => {
      // 0
      return t.binaryExpression(
        "+",
        nodes[index],
        t.binaryExpression(
          "+",
          t.stringLiteral(" "),
          nodes[index + 2] ? concat(nodes, index + 1) : nodes[index + 1]
        )
      );
    };

    const nodes = literals.length
      ? [
          t.stringLiteral(literals.map(literal => literal.value).join(" ")),
          ...identifiers,
        ]
      : [...identifiers];

    if (nodes.length === 1) {
      toAdd = nodes[0];
    } else if (nodes.length) {
      toAdd = concat(nodes, 0);
    } else {
      throw new MacroError(
        "ct.macro requires at least 1 argument and arguments must be literals or identifiers (variables)"
      );
    }

    argumentsNode.replaceWith(toAdd);
  });

  return {
    keepImports,
  };
});
