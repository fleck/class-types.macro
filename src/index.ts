export { default } from "./convertFunctionCallToStringMacro";

export { default as postcssClassTypes } from "./postcss";
