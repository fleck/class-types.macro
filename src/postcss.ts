import appRoot from "app-root-path";
import postcssTsClassnames from "postcss-ts-classnames/dist/plugin";
import nodePath from "path";
import fs from "fs-extra";

export const defaultDirectory = nodePath.join(
  appRoot.toString(),
  "@types",
  "class-types"
);

export const classnamesFilename = "classnames.d.ts";

export default ({ directory = "" } = {}) => {
  const dest = directory || defaultDirectory;

  fs.outputFile(
    nodePath.join(dest, "ct.macro.d.ts"),
    `\
// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
declare module "ct.macro" {
  const _default: (...args: ClassNames[]) => string;

  export default _default;
}`
  );

  return postcssTsClassnames({
    dest: nodePath.join(dest, classnamesFilename),
  });
};
