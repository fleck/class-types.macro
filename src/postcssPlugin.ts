import * as appRoot from "app-root-path";
import createSelectorParser from "postcss-selector-parser";
import nodePath from "path";
import fs from "fs-extra";
import postcssLib, { plugin } from "postcss";

export const defaultDirectory = nodePath.join(
  appRoot.toString(),
  "@types",
  "ct.macro"
);

export const classNamesDeclarationFilename = "classNames.d.ts";

const parser = createSelectorParser(selectors => {
  selectors.each(selector => {
    if (selector.type !== "selector") {
      selector.remove();
      return;
    }

    for (const node of selector.nodes) {
      if (node.type !== "class") {
        node.remove();
      }
    }
  });
});

const autoGeneratedFileWarning =
  "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n";

const initializer = ({ directory = "" } = {}) => {
  const dest = directory || defaultDirectory;

  const indexFileOutput = fs.outputFile(
    nodePath.join(dest, "index.d.ts"),
    `\
${autoGeneratedFileWarning}
declare module "ct.macro" {
const _default: (...args: ClassNames[]) => string;

export default _default;
}`
  );

  return async (root: postcssLib.Root) => {
    const classes: string[] = [];

    root.walkRules(rule => {
      classes.push(
        ...parser
          .processSync(rule.selector, {
            lossless: false,
            updateSelector: false,
          })
          .trim()
          .replace(".", "")
          .replace(/\\/g, "")
          .split(" ")
      );
    });

    return Promise.all([
      indexFileOutput,
      fs.outputFile(
        nodePath.join(dest, classNamesDeclarationFilename),
        `${autoGeneratedFileWarning}type ClassNames = "${classes.join(
          '" | "'
        )}"`
      ),
    ]);
  };
};

export type options = Parameters<typeof initializer>[0];

export const postcss = plugin<options>("postcss-ct.macro", initializer);
