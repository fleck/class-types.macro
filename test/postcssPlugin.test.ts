import postcss from "postcss";
import { postcss as plugin, options } from "../src/postcssPlugin";
import fs from "fs-extra";
import path from "path";

async function run(input: string, opts: options) {
  let result = await postcss([plugin(opts)]).process(input, {
    from: undefined,
  });

  expect(result.warnings()).toHaveLength(0);
}

const typesDirectory = path.join(__dirname, "..", "@types");

const cleanupFiles = () => {
  fs.removeSync(typesDirectory);
};

beforeEach(cleanupFiles);
afterEach(cleanupFiles);

const classNamesPath = path.join(typesDirectory, "ct.macro", "classNames.d.ts");

it("generates types for classes", async () => {
  await run(".mt-1{ } .mt-2{}", {});

  expect((await fs.readFile(classNamesPath)).toString()).toMatchSnapshot();
});

it("does not include ids, pseudo selectors, or attributes", async () => {
  await run(
    '#app .mt-1{ } #app .mt-2:hover{} [data-reach-slider-input][data-orientation="horizontal"] {}',
    {}
  );

  expect((await fs.readFile(classNamesPath)).toString()).toMatchSnapshot();
});

it("supports custom directories", async () => {
  const directory = path.join(typesDirectory, "custom");

  await run("#app .mt-1{ } #app .mt-2:hover{}", { directory });

  expect(
    (await fs.readFile(path.join(directory, "classNames.d.ts"))).toString()
  ).toMatchSnapshot();
});
