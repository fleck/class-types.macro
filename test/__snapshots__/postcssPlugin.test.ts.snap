// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not include ids, pseudo selectors, or attributes 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassNames = \\"mt-1\\" | \\"mt-2\\""
`;

exports[`generates types for classes 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassNames = \\"mt-1\\" | \\"mt-2\\""
`;

exports[`groups 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassNames = \\"group\\" | \\"group-hover:text-gray-900\\""
`;

exports[`supports custom directories 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
type ClassNames = \\"mt-1\\" | \\"mt-2\\""
`;
